OpenTrons/opentrons-api
# feat(step-generation): implement logic for collision detection with 96 partial pickup

Closes RAUT-1046

# Overview


This PR addresses a number of bugs to collision detection when using the 96-channel pipette with
single-column nozzle configuration. The main modifications lie in `getIsSafePipetteMovement`, which
detects the location of the pipette relative to the target labware, well, and offset, and confirms
whether any surrounding slots contain dangerous module + adapter + labware combinations.

# Test Plan

- create or upload a .json protocol to PD [example](https://github.com/user-attachments/files/16316808/Flex.LDH.96.well.kit.json)
- modify the starting deck state to introduce potential collisions and verify that high labware+module+adapter combos cause an error to be raised, and low labware does not cause issues

Some test cases that should raise a collision error:
- add a magnetic block with NEST deepwell plate directly to the left of a target aspiration or dispense
- add an Opentrons 15x 15ml Falcon tuberack directly to the left of a target aspiration or dispense

Some test cases that should not raise a collision error:
- add a temperature module with 24-tube aluminum block adapter directly to the left of a target aspiration or dispense
- add a 

# Changelog

- ensure tiprack and nozzles are passed as arguments to aspirate and dispense across complex command creators
- overhaul logic for pipette safety checks with 96-ch column configuration
- update tests

# Review requests

I added/modified some unit tests, please smoke test thoroughly. 

# Risk assessment

medium. a lot of step generation code is touched here, but it only implicates use of Flex 96-channel column configuration
